File: codegen/src/main/scala/bsp/codegen/docs/TypescriptRenderer.scala
package bsp.codegen.docs

import bsp.codegen.Lines
import bsp.codegen.common.ir
import bsp.codegen.dsl._
import bsp.codegen.common.ir.Def._
import bsp.codegen.common.ir.EnumType.{IntEnum, StringEnum}
import bsp.codegen.common.ir.Hint.{Deprecated, Documentation, Unstable}
import bsp.codegen.ir.Primitive._
import bsp.codegen.common.ir.Type._
import bsp.codegen.ir._
import cats.syntax.all._
import software.amazon.smithy.model.shapes.ShapeId

class TypescriptRenderer {
  import bsp.codegen.Settings.typescript

  def render(definition: ir.Def): Option[Lines] = {
    definition match {
      case Alias(shapeId, primitiveType, hints) =>
        Some(renderAlias(shapeId, primitiveType))
      case Structure(shapeId, fields, hints, associatedDataKinds) =>
        Some(renderStructure(shapeId, fields))
      case ClosedEnum(shapeId, enumType, values, hints) =>
        Some(renderClosedEnum(shapeId, enumType, values))
      case OpenEnum(shapeId, enumType, values, hints) =>
        Some(renderOpenEnum(shapeId, enumType, values))
      case Service(shapeId, operations, hints) => None
    }
  }

  def renderAlias(id: ShapeId, aliased: ir.Type): Lines = {
    val tpe = aliased match {
      case TPrimitive(prim, _) => renderPrimitive(prim)
      case _                   => renderType(aliased)
    }
    lines(
      s"export type ${id.getName()} = $tpe;"
    )
  }

  def renderStructure(shapeId: ShapeId, fields: List[ir.Field]): Lines = {
    lines(
      block(s"export interface ${shapeId.getName()}")(
        lines(fields.map(f => renderTSField(f)).intercalate(newline))
      )
    )
  }

  def renderClosedEnum[A](
      id: ShapeId,
      enumType: ir.EnumType[A],
      values: List[ir.EnumValue[A]]
  ): Lines = {
    val tpe = id.getName()
    lines(
      block(s"export enum $tpe") {
        values
          .map(value =>
            lines(
              renderDocumentation(value.hints),
              s"${renderStaticValue(enumType)(value)},"
            )
          )
          .intercalate(newline)
      }
    )
  }

  def renderEnumType[A](enumType: ir.EnumType[A]): String = {
    enumType match {
      case IntEnum    => renderPrimitive(Primitive.PInt)
      case StringEnum => renderPrimitive(Primitive.PString)
    }
  }

  def renderOpenEnum[A](
      shapeId: ShapeId,
      enumType: ir.EnumType[A],
      values: List[ir.EnumValue[A]]
  ): Lines = {
    val tpe = shapeId.getName()
    lines(
      s"export type $tpe = ${renderEnumType(enumType)};",
      newline,
      block(s"export namespace $tpe") {
        values
          .map(value =>
            lines(
              renderDocumentation(value.hints),
              s"export const ${renderStaticValue(enumType)(value)};"
            )
          )
          .intercalate(newline)
      }
    )
  }

  def renderStaticValue[A](enumType: ir.EnumType[A]): ir.EnumValue[A] => String = {
    enumType match {
      case IntEnum =>
        (ev: ir.EnumValue[Int]) => s"${camelCase(ev.name).capitalize} = ${ev.value}"

      case StringEnum =>
        (ev: ir.EnumValue[String]) => s"""${camelCase(ev.name).capitalize} = "${ev.value}""""
    }
  }

  def camelCase(string: String): String = {
    val first :: rest = string.split(Array(' ', '_')).toList.map(_.toLowerCase)
    val changedRest = rest.map(w => w.take(1).toUpperCase + w.drop(1))
    val reunited = first :: changedRest
    reunited.mkString
  }

  def renderEnumValueDef[A](enumType: ir.EnumType[A]): ir.EnumValue[A] => String = {
    enumType match {
      case IntEnum    => (ev: ir.EnumValue[Int]) => s"${ev.name}(${ev.value})"
      case StringEnum => (ev: ir.EnumValue[String]) => s"""${ev.name}("${ev.value}")"""
    }
  }

  def renderTSField(field: ir.Field): Lines = {
    val `:` = if (field.required) ": " else "?: "
    lines(
      renderDocumentation(field.hints),
      field.name + `:` + renderType(field.tpe) + ";"
    )
  }

  def renderType(tpe: ir.Type): String = tpe match {
    case TRef(shapeId) => shapeId.getName()
    case TPrimitive(prim, shapeId) =>
      if (shapeId.getNamespace == "smithy.api") {
        renderPrimitive(prim)
      } else {
        shapeId.getName
      }

    case TMap(key, value)     => s"Map<${renderType(key)}, ${renderType(value)}>"
    case TCollection(member)  => s"${renderType(member)}[]"
    case TSet(member)         => s"Set<${renderType(member)}>"
    case TUntaggedUnion(tpes) => tpes.map(renderType).mkString("|")
  }

  def renderPrimitive(prim: Primitive) = prim match {
    case PFloat     => "number"
    case PDouble    => "number"
    case PUnit      => "void"
    case PString    => "string"
    case PInt       => "number"
    case PDocument  => "any"
    case PBool      => "boolean"
    case PLong      => "number"
    case PTimestamp => "number"
  }

  def renderDocumentation(hints: List[ir.Hint]): Lines = {
    val lines = hints.flatMap {
      case Documentation(string) => string.split(System.lineSeparator())
      case Deprecated(message)   => Array(s"Deprecated: $message")
      case Unstable              => Array("Unstable: may change in the future")
    }.toArray

    if (lines.nonEmpty) {
      lines(0) = "/** " + lines(0)
      val lastIndex = lines.length - 1
      for (i <- 1 to lastIndex) {
        lines(i) = " * " + lines(i)
      }
      lines(lastIndex) = lines(lastIndex) + " */"
    }
    Lines(lines.toList)
  }
}

File: codegen/src/main/kotlin/bsp/codegen/docs/TypescriptRenderer.kt
package bsp.codegen.docs

import bsp.codegen.common.dsl.CodeBlock
import bsp.codegen.common.dsl.code
import bsp.codegen.common.ir.Def
import bsp.codegen.common.ir.EnumType
import bsp.codegen.common.ir.EnumValue
import bsp.codegen.common.ir.Field
import bsp.codegen.common.ir.Hint
import bsp.codegen.common.ir.Type
import bsp.codegen.common.util.snakeToUpperCamelCase
import kotlin.math.max

class TypescriptRenderer {
  fun render(definition: Def): CodeBlock? {
    return when (definition) {
      is Def.Alias -> renderAlias(definition)
      is Def.Structure -> renderStructure(definition)
      is Def.ClosedEnum<*> -> renderClosedEnum(definition)
      is Def.OpenEnum<*> -> renderOpenEnum(definition)
      is Def.Service -> throw RuntimeException("Service should not be rendered")
      is Def.DataKinds -> throw RuntimeException("DataKinds should not be rendered")
      is Def.UntaggedUnion -> throw RuntimeException("UntaggedUnion should not be rendered")
    }
  }

  fun renderAlias(definition: Def.Alias): CodeBlock = code {
    -"export type ${definition.shapeId.name} = ${renderType(definition.aliasedType)};"
  }

  private fun <T> List<T>.splitAtLast(): Pair<List<T>, List<T>> {
    val sizeWithoutLast = max(0, this.size - 1)
    return Pair(this.take(sizeWithoutLast), this.takeLast(1))
  }

  fun renderStructure(definition: Def.Structure): CodeBlock = code {
    val (withoutLast, last) = definition.fields.splitAtLast()
    block("export interface ${definition.shapeId.name}") {
      withoutLast.forEach {
        include(renderDocumentation(it.hints))
        -renderTSField(it)
        newline()
      }
      last.forEach {
        include(renderDocumentation(it.hints))
        -renderTSField(it)
      }
    }
  }

  fun renderClosedEnum(definition: Def.ClosedEnum<*>): CodeBlock = code {
    block("export enum ${definition.shapeId.name}") {
      definition.values.forEach {
        include(renderDocumentation(it.hints))
        -"${renderStaticValue(definition.enumType)(it)},"
        newline()
      }
    }
  }

  fun renderOpenEnum(definition: Def.OpenEnum<*>): CodeBlock = code {
    val enumType =
        renderType(
            when (definition.enumType) {
              is EnumType.IntEnum -> Type.Int
              is EnumType.StringEnum -> Type.String
            })
    -"export type ${definition.shapeId.name} = $enumType;"
    newline()
    block("export namespace ${definition.shapeId.name}") {
      definition.values.forEach {
        include(renderDocumentation(it.hints))
        -"export const ${renderStaticValue(definition.enumType)(it)};"
        newline()
      }
    }
  }

  fun renderStaticValue(enumType: EnumType<*>): (EnumValue<*>) -> String = { value ->
    when (enumType) {
      is EnumType.IntEnum -> "${value.name.snakeToUpperCamelCase()} = ${value.value}"
      is EnumType.StringEnum -> "${value.name.snakeToUpperCamelCase()} = \"${value.value}\""
    }
  }

  fun renderTSField(field: Field): String =
      "${field.name}${"?".takeUnless { field.required } ?: ""}: ${renderType(field.type)};"

  fun renderType(tpe: Type): String =
      when (tpe) {
        is Type.Ref -> tpe.shapeId.name
        Type.Bool -> "boolean"
        Type.Int -> "number"
        Type.Json -> "any"
        is Type.List -> "${renderType(tpe.member)}[]"
        Type.Long -> "number"
        is Type.Map -> "Map<${renderType(tpe.key)}, ${renderType(tpe.value)}>"
        is Type.Set -> "Set<${renderType(tpe.member)}>"
        Type.String -> "string"
        Type.Unit -> "void"
        is Type.UntaggedUnion -> tpe.members.joinToString(" | ") { renderType(it) }
      }

  fun renderDocumentation(hints: List<Hint>): CodeBlock = code {
    val lines =
        hints.flatMap {
          when (it) {
            is Hint.Documentation -> it.string.split(System.lineSeparator())
            is Hint.Deprecated -> listOf("Deprecated: ${it.message}")
            Hint.Unstable -> listOf("Unstable: may change in the future")
            is Hint.JsonRename -> listOf()
          }
        }

    val commentLines = lines.toMutableList()

    if (commentLines.isNotEmpty()) {
      commentLines[0] = "/** ${commentLines[0]}"
      val lastIndex = commentLines.size - 1
      for (i in 1..lastIndex) {
        commentLines[i] = " * ${commentLines[i]}"
      }
      commentLines[lastIndex] = "${commentLines[lastIndex]} */"
    }

    commentLines.forEach { -it }
  }
}
